tests/msc3083_test.go

```diff
@@ -181,6 +184,275 @@ func TestRestrictedRoomsRemoteJoin(t *testing.T) {
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+
+			return true
+		},
+	)
+
+	// Raise the power level so that users on hs1 can invite people and then leave
+	// the room.
+	state_key := ""
+	charlie.SendEventSynced(t, room, b.Event{
+		Type:     "m.room.power_levels",
+		StateKey: &state_key,
+		Content: map[string]interface{}{
+			"invite": 0,
+			"users": map[string]interface{}{
+				charlie.UserID: 100,
+			},
+		},
+	})
+	charlie.LeaveRoom(t, room)
+
+	// Ensure the events have synced to hs1.
+	alice.SyncUntilTimelineHas(
+		t,
+		room,
+		func(ev gjson.Result) bool {
+			if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != charlie.UserID {
+				return false
+			}
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "leave", "Charlie failed to leave the room")
+
+			return true
+		},
+	)
+
+	// Have bob leave and rejoin. This should still work even though hs2 isn't in
+	// the room anymore!
+	bob.LeaveRoom(t, room)
+	bob.JoinRoom(t, room, []string{"hs1"})
+}
+
+// A server will request a failover if asked to /make_join and it does not have
+// the appropriate authorisation to complete the request.
+//
+// Setup 3 homeservers:
+// * hs1 creates the space/room.
+// * hs2 joins the room
+// * hs3 attempts to join via hs2 (should fail) and hs3 (should work)
+func TestRestrictedRoomsRemoteJoinFailOver(t *testing.T) {
+	deployment := Deploy(t, b.Blueprint{
+		Name: "federation_three_homeservers",
+		Homeservers: []b.Homeserver{
+			{
+				Name: "hs1",
+				Users: []b.User{
+					{
+						Localpart:   "alice",
+						DisplayName: "Alice",
+					},
+				},
+			},
+			{
+				Name: "hs2",
+				Users: []b.User{
+					{
+						Localpart:   "bob",
+						DisplayName: "Bob",
+					},
+				},
+			},
+			{
+				Name: "hs3",
+				Users: []b.User{
+					{
+						Localpart:   "charlie",
+						DisplayName: "Charlie",
+					},
+				},
+			},
+		},
+	})
+	defer deployment.Destroy(t)
+
+	// Setup the user, space, and restricted room.
+	alice, space, room := setupRestrictedRoom(t, deployment)
+
+	// Raise the power level so that only alice can invite.
+	state_key := ""
+	alice.SendEventSynced(t, room, b.Event{
+		Type:     "m.room.power_levels",
+		StateKey: &state_key,
+		Content: map[string]interface{}{
+			"invite": 100,
+			"users": map[string]interface{}{
+				alice.UserID: 100,
+			},
+		},
+	})
+
+	// Create a second user on a different homeserver.
+	bob := deployment.Client(t, "hs2", "@bob:hs2")
+
+	// Bob joins the room and space.
+	bob.JoinRoom(t, space, []string{"hs1"})
+	bob.JoinRoom(t, room, []string{"hs1"})
+
+	// Charlie should join the space (which gives access to the room).
+	charlie := deployment.Client(t, "hs3", "@charlie:hs3")
+	charlie.JoinRoom(t, space, []string{"hs1"})
+
+	// hs2 doesn't have anyone to invite from, so the join fails.
+	failJoinRoom(t, charlie, room, "hs2", 502)
```

Comment:
Fairly sure there won't be any guarantee that hs2 would've seen the join from hs3 because the join was between hs3 and hs1. Therefore, doing this will be racey.